frontend/javascripts/test/sagas/volumetracing_saga_integration.spec.js

```diff
@@ -64,14 +84,223 @@ test.serial(
+  Store.dispatch(addToLayerAction(paintCenter));
+  Store.dispatch(finishEditingAction());
+
+  await sleep(2000);
+
+  Store.dispatch(undoAction());
+
+  t.is(await t.context.api.data.getDataValue("segmentation", paintCenter), newCellId);
+  t.is(await t.context.api.data.getDataValue("segmentation", [1, 0, 0]), newCellId);
+  t.is(await t.context.api.data.getDataValue("segmentation", [5, 0, 0]), oldCellId);
+});
+
+test.serial("Brushing/Tracing with undo (II)", async t => {
+  const oldCellId = 11;
+  t.context.mocks.Request.sendJSONReceiveArraybufferWithHeaders = createBucketResponseFunction(
+    oldCellId,
+    500,
+  );
+
+  const paintCenter = [0, 0, 0];
+  const brushSize = 10;
+
+  const newCellId = 2;
+
+  Store.dispatch(updateUserSettingAction("brushSize", brushSize));
+  Store.dispatch(setPositionAction([0, 0, 0]));
+  Store.dispatch(setToolAction("BRUSH"));
+
+  Store.dispatch(setActiveCellAction(newCellId));
+  Store.dispatch(startEditingAction(paintCenter, OrthoViews.PLANE_XY));
+  Store.dispatch(addToLayerAction(paintCenter));
+  Store.dispatch(finishEditingAction());
+
+  Store.dispatch(setActiveCellAction(newCellId + 1));
+  Store.dispatch(startEditingAction(paintCenter, OrthoViews.PLANE_XY));
+  Store.dispatch(addToLayerAction(paintCenter));
+  Store.dispatch(finishEditingAction());
+
+  Store.dispatch(setActiveCellAction(newCellId + 2));
+  Store.dispatch(startEditingAction(paintCenter, OrthoViews.PLANE_XY));
+  Store.dispatch(addToLayerAction(paintCenter));
+  Store.dispatch(finishEditingAction());
+
+  await sleep(2000);
+
+  Store.dispatch(undoAction());
+
+  t.is(await t.context.api.data.getDataValue("segmentation", paintCenter), newCellId + 1);
+  t.is(await t.context.api.data.getDataValue("segmentation", [1, 0, 0]), newCellId + 1);
+  t.is(await t.context.api.data.getDataValue("segmentation", [5, 0, 0]), oldCellId);
+});
+
+test.serial(
+  "Brushing/Tracing should crash when too many buckets are labeled at once without saving inbetween",
+  async t => {
+    // webKnossos will start to evict buckets forcefully if too many are dirty at the same time.
+    // This is not ideal, but usually handled by the fact that buckets are regularly saved to the
+    // backend and then marked as not dirty.
+    // This test provokes that webKnossos crashes (a hard crash is only done during testing; in dev/prod
+    // a soft warning is emitted via the devtools).
+    // The corresponding sibling test checks that saving inbetween does not make webKnossos crash.
+    t.plan(1);
+    const failedSagaPromise = waitForCondition(hasRootSagaCrashed, 500);
+    await Promise.race([testLabelingManyBuckets(t, false), failedSagaPromise]);
+    t.true(hasRootSagaCrashed());
+  },
+);
+
+test.serial(
+  "Brushing/Tracing should send buckets to backend and restore dirty flag afterwards",
+  async t => {
+    t.plan(1);
+    await testLabelingManyBuckets(t, true);
+    // If nothing crashed, the test was successful.
+    t.true(true);
```

Comment:
Wouldn't you need to test `t.false(hasRootSagaCrashed());` as the error is catched in the `root_saga`?

Reply:
Good point!